Use asyncio to speed up cloud identification

kshivakumar opened the pull request:
Implements https://github.com/dgzlopes/cloud-detect/issues/11

Uses `asyncio` to check all the cloud providers concurrently and returns back a response as soon as the cloud provider is detected.  

**Important changes:**
 - Replaced `requests` with `aiohttp` in the dependencies
 - Removed Python 3.4 and 3.5 support.
 - Added `identifier` class variable to each `*Provider` class
 - Added `SUPPORTED_PROVIDERS` api - it's a tuple of `identifier` values of each Provider class
 - Added support for Python 3.9 - the code and test cases are working

I manually tested the changes in Azure and GCP, it's working fine and the response comes back within a second.

kshivakumar commented:
For some unknown reason, I get the below error during pre-commit/make test:
```
Reorder python imports...................................................Failed
- hook id: reorder-python-imports
- exit code: 1

Traceback (most recent call last):
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/bin/reorder-python-imports", line 8, in <module>
    sys.exit(main())
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/lib/python3.7/site-packages/reorder_python_imports.py", line 685, in main
    application_directories=args.application_directories.split(':'),
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/lib/python3.7/site-packages/reorder_python_imports.py", line 420, in fix_file_contents
    partitioned = step(partitioned)
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/lib/python3.7/site-packages/reorder_python_imports.py", line 334, in apply_import_sorting
    sorted_blocks = sort(import_obj_to_partition.keys(), **sort_kwargs)
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/lib/python3.7/site-packages/aspy/refactor_imports/sort.py", line 96, in sort
    imports_partitioned[classify_func(import_obj)].append(import_obj)
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/lib/python3.7/site-packages/aspy/refactor_imports/sort.py", line 73, in classify_func
    return classify_import(obj.import_statement.module, **kwargs)
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/lib/python3.7/site-packages/aspy/refactor_imports/classify.py", line 139, in classify_import
    base, application_directories,
  File "/root/.cache/pre-commit/repol5aizj_t/py_env-python3/lib/python3.7/site-packages/aspy/refactor_imports/classify.py", line 103, in _get_module_info
    assert spec.submodule_search_locations is not None
AssertionError
```
I tried different Python versions (3.7, 3.8, 3.9), but the error doesn't go away.

kshivakumar commented on cloud_detect/__init__.py:
```diff
@@ -7,26 +10,65 @@
+
+    if tasks:
+        await cancel_unfinished_tasks()
+        return 'timeout'
```
If the code is run in a non-supported cloud environment or in local, the response is usually going to be`timeout` because almost all the cloud provider urls take more than 5s to respond. I am wondering if we should return `unknown in this case?

kshivakumar replied:
In my latest commits I have set `timeout`'s default value as `None` and removed "timeout" as a response.
So, by default, `provider` will take all its time to determine the cloud provider. In my local, it took around 1 min 15s to return `unknown`.

kshivakumar commented on cloud_detect/__init__.py:
```diff
@@ -7,26 +10,65 @@
+    return result
+
+
+async def async_provider(timeout=TIMEOUT):
```
If the user code is async, then they should use this api instead of `provider`.

kshivakumar commented:
@dgzlopes Test cases are failing for Python 3.5 because the library for mocking api responses - [`aresponses`](https://github.com/aresponses/aresponses) doesn't support Python 3.5.
I have to check if there's an alternative library that supports Python 3.5.

kshivakumar commented:
@dgzlopes How important is Python 3.5 support? It's officially retired - https://endoflife.date/python.

dgzlopes commented:
Not important @kshivakumar :D

Feel free to drop it!

dgzlopes commented:
I'll review the PR at the end of the week (I've scheduled it on my personal calendar).

Sorry for my slow pace :smile:

dgzlopes reviewed (approved):
Thanks again for the PR! LGTM :) 

The only thing, while running tox locally, the pre-commit hook step did some fixes on `setup.py` (re-ordering Python import). Do you mind committing those?

Thanks again!

kshivakumar commented:
> The only thing, while running tox locally, the pre-commit hook step did some fixes on `setup.py` (re-ordering Python import). Do you mind committing those?
> 

Fixed it and also updated README.

kshivakumar commented:
@dennis-pg Can we merge this?

kshivakumar commented:
@dgzlopes Can we merge this if there are no further issues?

dgzlopes commented:
Sorry for the delay! I had a mess on my GH notifications :( 

Merged. I'll cut a new release.
